src/sentry/hybridcloud/models/outbox.py

```diff
@@ -333,21 +468,45 @@ def drain_shard(
                 return
 
         shard_row: OutboxBase | None
-        while True:
-            with self.process_shard(latest_shard_row) as shard_row:
-                if shard_row is None:
-                    break
+        reservation_shards = options.get("hybrid_cloud.outbox.reservation_shards")
+        if self.shard_scope in reservation_shards:
+            # Reservation shard logic does not fire signal handlers within a database transaction
+            while True:
+                # Reserve and fetch messages from the shard until it is empty
+                with self.reserve_coalesced_messages(
+                    latest_shard_row=latest_shard_row,
+                    is_synchronous_flush=not flush_all,
+                ) as next_outbox:
+                    if next_outbox is None:
+                        break
 
-                if _test_processing_barrier:
-                    _test_processing_barrier.wait()
+                    if _test_processing_barrier:
+                        _test_processing_barrier.wait()
 
-                processed = shard_row.process(is_synchronous_flush=not flush_all)
+                    processed = next_outbox.process_signal(is_synchronous_flush=not flush_all)
 
-                if _test_processing_barrier:
-                    _test_processing_barrier.wait()
+                    if _test_processing_barrier:
+                        _test_processing_barrier.wait()
 
-                if not processed:
-                    break
+                    if not processed:
+                        break
+        else:
+            while True:
+                with self.process_shard(latest_shard_row) as shard_row:
+                    # All of this logic is in the select_for_update block, with row locks
+                    if shard_row is None:
+                        break
+
+                    if _test_processing_barrier:
+                        _test_processing_barrier.wait()
+
+                    processed = shard_row.process(is_synchronous_flush=not flush_all)
+
+                    if _test_processing_barrier:
+                        _test_processing_barrier.wait()
+
+                    if not processed:
+                        break
```

Comment:
This logic is unchanged, just indented to be inside the option guard. It is worth noting that this entire block happens within the locking transaction, which is part of the problem.  I think this lock was retained in #84631 so I'm not sure that it would have fixed the statement timeout problem.